src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10006,7 +10012,7 @@ private bool ScanParenthesizedImplicitlyTypedLambda(Precedence precedence)
             //  case 1:  ( x ,
             if (this.PeekToken(1).Kind == SyntaxKind.IdentifierToken
                 && (!this.IsInQuery || !IsTokenQueryContextualKeyword(this.PeekToken(1)))
-                && this.PeekToken(2).Kind == SyntaxKind.CommaToken)
+                && (this.PeekToken(2).Kind == SyntaxKind.CommaToken || (this.PeekToken(2).Kind == SyntaxKind.ExclamationToken && this.PeekToken(3).Kind == SyntaxKind.CommaToken)))
```

Comment:
these sorts of complex conditionals just get harder and harder to parse.  can you just extract out individual checks (or a helper function)?  Then, the code can clearly enumerate the cases it cares about, including understandable comments explaining what's going on.  Bonus: you can have temp variables so we're not constantly having to peek to the same tokesn over and over again.

Reply:
Should I extract out a function that determines which case it falls under (or none), or individual functions for each case?

Reply:
i'd be fine with a single static local function that determines true/false, but doens't feel the need to try to smoosh every case into one large expression.  Tnx!